From 631e7b2bb302030d2879255e0a263dfe3f772a5b Mon Sep 17 00:00:00 2001
From: pramod kotreshappa <kpramod@codeaurora.org>
Date: Thu, 9 May 2019 15:52:05 -0700
Subject: [PATCH 21/37] TWS_A2DP: Handle active device change between TWS+
 earbuds

Check for TWS+ device switch in active device change. If
TWS+ device switch then remove disconnected device
and add peer pair device into connected devices list and
return.

[wight554: updated code from LA.QSSI.11.0.r1-05600-qssi.0]

CRs-fixed: 2446639
Change-Id: I97e66fba56fceb2e9b2e7c253119033bf34175bb
Signed-off-by: Volodymyr Zhdanov <wight554@gmail.com>
---
 .../server/audio/AudioDeviceInventory.java        |  6 ++++++
 .../java/com/android/server/audio/BtHelper.java   | 15 +++++++++++++++
 2 files changed, 21 insertions(+)

diff --git a/services/core/java/com/android/server/audio/AudioDeviceInventory.java b/services/core/java/com/android/server/audio/AudioDeviceInventory.java
index 1b26216322e..43ef518306b 100644
--- a/services/core/java/com/android/server/audio/AudioDeviceInventory.java
+++ b/services/core/java/com/android/server/audio/AudioDeviceInventory.java
@@ -869,6 +869,12 @@ public class AudioDeviceInventory {
                       mConnectedDevices.put(deviceKey, new DeviceInfo(
                                  AudioSystem.DEVICE_OUT_BLUETOOTH_A2DP, BtHelper.getName(device),
                                  address, a2dpCodec));
+                      if (BtHelper.isTwsPlusSwitch(device, existingDevice.getValue().mDeviceAddress)) {
+                          if (AudioService.DEBUG_DEVICES) {
+                              Log.d(TAG,"TWS+ device switch");
+                          }
+                          return;
+                      }
                       mDeviceBroker.postA2dpActiveDeviceChange(
                                  new BtHelper.BluetoothA2dpDeviceInfo(
                                      device, a2dpVolume, a2dpCodec));
diff --git a/services/core/java/com/android/server/audio/BtHelper.java b/services/core/java/com/android/server/audio/BtHelper.java
index b4c41b274db..c360980f584 100644
--- a/services/core/java/com/android/server/audio/BtHelper.java
+++ b/services/core/java/com/android/server/audio/BtHelper.java
@@ -219,6 +219,21 @@ public class BtHelper {
         return deviceName;
     }
 
+    /*packages*/ @NonNull static boolean isTwsPlusSwitch(@NonNull BluetoothDevice device,
+                                                                 String address) {
+        BluetoothAdapter adapter = BluetoothAdapter.getDefaultAdapter();
+        if (device == null || adapter.getRemoteDevice(address) == null ||
+            device.getTwsPlusPeerAddress() == null) {
+            return false;
+        }
+        if (device.isTwsPlusDevice() &&
+            adapter.getRemoteDevice(address).isTwsPlusDevice() &&
+            device.getTwsPlusPeerAddress().equals(address)) {
+            Log.i(TAG,"isTwsPlusSwitch true");
+            return true;
+         }
+         return false;
+    }
     //----------------------------------------------------------------------
     // Interface for AudioDeviceBroker
 
-- 
2.28.0

